### PR TITLE
[CORE-214] - implement bbDisplayMode as ViewportSize service

### PR DESCRIPTION
**Pull request by TolgaCura**

### CODE REVIEW

File: src/core/javascripts/services/viewport_size.js.coffee [hunk 1]
```diff
@@ -7,15 +7,93 @@
...
-  setViewportSize: (size) ->
-    if size != viewport_size
-      viewport_size = size
-      $rootScope.$broadcast 'ViewportSize:changed'
+  # id prefix for span html elements used to determin screen size via bootstrap classes
+  viewport_element_id_prefix = 'viewport_size_'
 
-  getViewportSize: () ->
+  # using function to grab screensize so it cannot be altered outside service
+  service.getViewportSize = ->
```
Comment (lukeferguson) [inline src/core/javascripts/services/viewport_size.js.coffee]: Using the revealing module pattern to expose methods - see the i18 service for an example

File: src/core/javascripts/services/viewport_size.js.coffee [hunk 2]
```diff
@@ -7,15 +7,93 @@
...
+
+  # determins the current size of the screen
+  do calculateCurrentSize = ->
+    viewport_elements = getViewportElementsFromDocumentBody()
+    for viewport_element in viewport_elements
+      element_size = getSizeFromElement(viewport_element)
+      if isElementVisible(viewport_element)
+        viewport_size = element_size
+        # create function to return true for current screensize 
+        service['is' + element_size.toUpperCase()] = -> return true
```
Comment (lukeferguson) [inline src/core/javascripts/services/viewport_size.js.coffee]: Hmmm, dynamically creating methods will make use the revealing module pattern hard - maybe stick to ordering public methods towards the bottom of the file for this one

File: src/core/javascripts/controllers/base.js.coffee [hunk 3]
```diff
@@ -104,7 +104,7 @@ BBCtrl = ($scope, $location, $rootScope, halClient, $window, $http, $q, $timeout
 
   $onInit = () ->
 
-    compileDisplayMode()
+    ViewportSize.init() # Initialise viewport size tracking
```
Comment (damianpudlo) [inline src/core/javascripts/controllers/base.js.coffee]: we should move it inside vm.$postLink lifecycle hook

File: src/core/javascripts/services/viewport_size.js.coffee [hunk 4]
```diff
@@ -7,15 +7,159 @@
...
+  # @description determins the current size of the screen and generates dynamic functions
+  ###
+  findVisibleElement = ->
+    viewport_elements = getViewportElementsFromDocumentBody()
+    for viewport_element in viewport_elements
+      element_size = getSizeFromElement(viewport_element)
+      if isElementVisible(viewport_element)
+        viewport_size = element_size
+        # dynamically created function to return true for current screensize 
+        service['is' + element_size.toUpperCase()] = -> return true
```
Comment (damianpudlo) [inline src/core/javascripts/services/viewport_size.js.coffee]: can we make it more explicit by

1)  having state object on the top 
like
 state = {
    isXS: false
    isSM: false
    isMD: false
    isLG: false
  }

2) changing: 
service['is' + element_size.toUpperCase()] = -> return true 
service['is' + element_size.toUpperCase()] = -> return false
appropriately to 
state['is' + element_size.toUpperCase()] = true
state['is' + element_size.toUpperCase()] = false



3) defining get functions 
service.isXS = () ->
  return stat.isXS
...
...
...

File: src/core/javascripts/services/viewport_size.js.coffee [hunk 5]
```diff
@@ -7,15 +7,159 @@
...
   viewport_size = null
 
-  setViewportSize: (size) ->
-    if size != viewport_size
-      viewport_size = size
-      $rootScope.$broadcast 'ViewportSize:changed'
+  ###
+  # @description id prefix for span html elements used to determin screen size via bootstrap classes
+  ###
+  viewport_element_id_prefix = 'viewport_size_'
```
Comment (damianpudlo) [inline src/core/javascripts/services/viewport_size.js.coffee]: styleguides changed last months and we switched to standard camelCase for all new vars and properties

**Pull request merged**